### ISSUE TITLE
Add tests to `universum`

### ISSUE THREAD

**Issue by ChShersh** [serokell/universum#36]
We mostly reexport some basic stuff. But I think some functions can and should be tested on correctness. Here is list of some functions:[
`encodeUtf8`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Conv.hs#L29) [`decodeUtf8`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Conv.hs#L30) [`ordNub`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/List.hs#L37) [`unsnoc`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Universum.hs#L214) [`andM`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Monad.hs#L99) [`orM`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Monad.hs#L105) [`anyM`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Monad.hs#L111) [`allM`](https://github.com/serokell/universum/blob/a48e2af3e618412156e4b2996dcd1d61133ab023/src/Monad.hs#L117)

**Comment by johnchandlerburnham** [comment]
I've added some property tests with tasty and hedgehog on my fork here: https://github.com/johnchandlerburnham/universum/commit/3a610bbd348a3504be0f97f94f2e439426b57bd9

Can send a pull request, but wanted to collaborate a little bit first on what the testing infrastructure for universum should look like.

**Comment by johnchandlerburnham** [comment]
Also, the utf8 tests in my fork fail on the string `"\65535"`, which I've tracked down to  `Data.ByteString.UTF8`:

```
> import qualified Data.ByteString.UTF8 as BU                  
BU> BU.toString (BU.fromString "\65534") == "\65533"          
True
BU> BU.toString (BU.fromString "\65535") == "\65533"          
True
```

I think this is just a mismatch between what [Data.ByteString.UTF8.fromString](https://hackage.haskell.org/package/utf8-string-1.0.1.1/docs/src/Data-ByteString-UTF8.html#toString) thinks is valid and what [Test.Hedgehog.Gen.unicode]( http://hackage.haskell.org/package/hedgehog-0.5.2/docs/src/Hedgehog-Internal-Gen.html#unicode) thinks is valid.

**Comment by ChShersh** [comment]
@johnchandlerburnham Thanks a lot for your interest and contribution to `universum`! I really appreciate that 🙂 Feel free to open PR. I will review and left my comments. At first glance: your work is great but I think it can be slightly improved 🙂  I also want `tasty` and `hedgehog` be used so it's really nice that you chose these libraries!

Any particular reason why you used `tasty-0.12` and set `allow-newer: true`? I would like to avoid putting something to `extra-deps` unless there's actual reason for this (because it stops `universum` from releasing to Stackage).

> Also, the utf8 tests in my fork fail on the string "\65535"

Argh, there're some problems with unicode in recent times... `doctest` become broken after bump up to `text-1.2.3.` (see this PR #127), also `text` has some non-fixed problems at them moment (https://github.com/haskell/text/pull/212), etc.

> I think this is just a mismatch between what Data.ByteString.UTF8.fromString thinks is valid and what Test.Hedgehog.Gen.unicode thinks is valid.

I think you can open issue in `hedgehog` library (referencing this issue) so maintainers of this library can guide us regarding correct and expected behavior:

* https://github.com/hedgehogqa/haskell-hedgehog/issues

**Comment by johnchandlerburnham** [comment]
Okay, opened a pull request at https://github.com/hedgehogqa/haskell-hedgehog/issues/153

No particular reason for using `tasty-0.12.0.1`, I changed things to `tasty-11.3`. 

Before I do a PR I had a question about  the right way to handle nix integration with stack. I'm on NixOS, so I enabled nix integration in `stack.yaml` and moved `binutils`, `gmp` to a `shell.nix` file. I don't want to break things for people who have stack but not nix, so should I just put

```
nix:
  shell-file: shell.nix
```
in the `stack.yaml` and then nix users can build by passing `--nix` flag to stack? I'm just wondering if there's  a better way to handle this.
